src/v/kafka/server/connection_context.h

```diff
@@ -106,7 +106,9 @@ class connection_context final
 
     protocol& server() { return _proto; }
     const ss::sstring& listener() const { return _rs.conn->name(); }
-    security::sasl_server& sasl() { return _sasl; }
+    security::sasl_server* sasl() {
```

Comment:
Would be nice to avoid returning maybe-null pointers: I think this function can assert on sasl being set and always return a reference, if we change the checks in connection_context.cc to look at whether _sasl is set rather than whether this function returns null.

That way all the other places we call sasl() in sasl_authenticate can be safe, without having to carry through the (correct, but perhaps fragile) assumption that once we've reached that code this pointer must be non-null.

Reply:
But there are places where the nullness of sasl is checked to determine whether sasl is enabled. before dereferencing it and crashing. I could represent the optionality of sasl with `std::optional<std::reference_wrapper<sasl_server>>`, if that's less objectionable?

Reply:
I think (?) that the explicit checks were only in connection_context.cc, where it can see _sasl directly without using the getter.  Then in other locations I think it was sasl-specific code that was skipping the check anyway, so this getter could assert the optional is set before returning a `sasl_server&`

(This does not hold if there were places outside of connection_context.cc that needed that explicit check)

Reply:
There are places outside connection_context. I'm changing this to return `optional<sasl_server>&`.